Using Synth Volumes with Simulation crashes

Tried to use our new volume.  The following code fails with incorrect dtypes.

```
...
        self.sim = Simulation(
            vols=AsymmetricVolume(32,1).generate(),
            n=1024,
            offsets=0,
        )
``` 

```
        if self.vols.dtype != self.dtype:
>           raise RuntimeError(
                f"{self.__class__.__name__} dtype {self.dtype}"
                f" does not match provided vols.dtype {self.vols.dtype}."
            )
E           RuntimeError: Simulation dtype float32 does not match provided vols.dtype float64.
```

I believe we want to pass through here. Will push a patch.

--- a/src/aspire/source/simulation.py
+++ b/src/aspire/source/simulation.py
@@ -34,7 +34,7 @@ class Simulation(ImageSource):
         filter_indices=None,
         offsets=None,
         amplitudes=None,
-        dtype=np.float32,
+        dtype=None,
         C=2,
         angles=None,
         seed=0,
@@ -47,30 +47,36 @@ class Simulation(ImageSource):
 
         :param angles: A n-by-3 array of rotation angles
         """
-        super().__init__(L=L, n=n, dtype=dtype, memory=memory)
 
         self.seed = seed
 
         # If a Volume is not provided we default to the legacy Gaussian blob volume.
-        # If a Simulation resolution is not provided, we default to L=8.
+        # If a Simulation resolution or dtype is not provided, we default to L=8 and np.float32.
         if vols is None:
             self.vols = LegacyVolume(
-                L=L or 8, C=2, seed=self.seed, dtype=self.dtype
+                L=L or 8,
+                C=2,
+                seed=self.seed,
+                dtype=dtype or np.float32,
             ).generate()
         else:
+            if dtype is not None and vols.dtype != dtype:
+                raise RuntimeError(
+                    f"Explicit {self.__class__.__name__} dtype {dtype}"
+                    f" does not match provided vols.dtype {vols.dtype}."
+                    " Please change the Volume using `astype`"
+                    " or the Simuation `dtype` argument."
+                )
+            # Assign the explicitly provided volume.
             self.vols = vols
 
         if not isinstance(self.vols, Volume):
             raise RuntimeError("`vols` should be a Volume instance or `None`.")
 
-        if self.vols.dtype != self.dtype:
-            raise RuntimeError(
-                f"{self.__class__.__name__} dtype {self.dtype}"
-                f" does not match provided vols.dtype {self.vols.dtype}."
-            )
-            self.vols = self.vols.astype(self.dtype)
-
-        self.L = self.vols.resolution
+        # Infer the details from volume when possible.
+        super().__init__(
+            L=self.vols.resolution, n=n, dtype=self.vols.dtype, memory=memory
+        )
 
         # If a user provides both `L` and `vols`, resolution should match.
         if L is not None and L != self.L:

--- a/src/aspire/volume/volume.py
+++ b/src/aspire/volume/volume.py
@@ -87,14 +87,16 @@ class Volume:
         """
         return self._data
 
-    def astype(self, dtype):
+    def astype(self, dtype, copy=True):
         """
         Return `Volume` instance with the prescribed dtype.
 
         :param dtype: Numpy dtype
+        :param copy: Boolean, optionally avoid copying if Volume.dtype already matches.
+            Defaults to True.
         :return: Volume instance
         """
-        return Volume(self.asnumpy().astype(dtype))
+        return Volume(self.asnumpy().astype(dtype, copy=copy))
 
     def __getitem__(self, item):
         # this is one reason why you might want Volume and VolumeStack classes...

--- a/tests/test_simulation.py
+++ b/tests/test_simulation.py
@@ -50,6 +50,41 @@ class SimVolTestCase(TestCase):
         with raises(RuntimeError, match=r"`vols` should be a Volume instance*"):
             _ = Simulation(L=self.vol_res, vols=self.vol_arr)
 
+    def testDtypeMismatch(self):
+        """
+        Test we raise when the volume dtype does not match explicit Simulation dtype.
+        """
+        with raises(RuntimeError, match=r".*does not match provided vols.dtype.*"):
+            _ = Simulation(vols=self.vol.astype(np.float16), dtype=self.dtype)
+
+    def testPassthroughFromVol(self):
+        """
+        Test we do not crash when passing a volume to Simulation,
+        without an explcit Simulation dtype.
+        """
+        for dtype in (np.float32, np.float64):
+            sim = Simulation(vols=self.vol.astype(dtype, copy=False))
+            # Did we assign the right type?
+            self.assertTrue(sim.dtype == dtype)
+
+            # Is the Volume the intended type?
+            self.assertTrue(sim.vols.dtype == dtype)
+
+    def testPassthroughFromSim(self):
+        """
+        Test we do not crash when passing a volume to Simulation,
+        with out an explcit Volume.
+        """
+        for dtype in (np.float32, np.float64):
+            # Create a minimal sim
+            sim = Simulation(dtype=dtype)
+
+            # Did we assign the right type?
+            self.assertTrue(sim.dtype == dtype)
+
+            # Is the Volume the intended type?
+            self.assertTrue(sim.vols.dtype == dtype)
+
 
 class SimTestCase(TestCase):
     def setUp(self):

--- a/tests/test_volume.py
+++ b/tests/test_volume.py
@@ -53,6 +53,28 @@ class VolumeTestCase(TestCase):
         self.assertTrue(np.allclose(v2.asnumpy(), self.vols_1.asnumpy()))
         self.assertTrue(v2.dtype == new_dtype)
 
+    def testAsTypeCopy(self):
+        """
+        `astype(copy=False)` is an optimization partially mimicked from numpy.
+        """
+        # Same dtype, copy=False
+        v2 = self.vols_1.astype(self.vols_1.dtype, copy=False)
+        # Details should match,
+        self.assertTrue(isinstance(v2, Volume))
+        self.assertTrue(np.allclose(v2.asnumpy(), self.vols_1.asnumpy()))
+        self.assertTrue(v2.dtype == self.vols_1.dtype)
+        # and they should share the same memory (np.ndarray.base).
+        self.assertTrue(v2.asnumpy().base is self.vols_1.asnumpy().base)
+
+        # Same dtype, default copy=True
+        v2 = self.vols_1.astype(self.vols_1.dtype)
+        # Details should match,
+        self.assertTrue(isinstance(v2, Volume))
+        self.assertTrue(np.allclose(v2.asnumpy(), self.vols_1.asnumpy()))
+        self.assertTrue(v2.dtype == self.vols_1.dtype)
+        # but they should not share the same memory (np.ndarray.base)
+        self.assertTrue(v2.asnumpy().base is not self.vols_1.asnumpy().base)
+
     def testGetter(self):
         k = np.random.randint(self.n)
         self.assertTrue(np.all(self.vols_1[k] == self.data_1[k]))

